cfy_manager/components/postgresql_server/postgresql_server.py

```diff
@@ -377,6 +386,257 @@ def _configure_cluster(self):
+                else:
+                    replicas.append(server_name)
+        else:
+            raise DBNodeListError(
+                'Can only list DB nodes from a manager or DB node.'
+            )
+        return master, replicas
+
+    def _get_node_status(self, address, target_type):
+        url = {
+            'etcd': 'https://{address}:2379/v2/stats/self',
+            'DB': 'https://{address}:8008',
+        }[target_type].format(address=address)
+
+        dead_node_exceptions = (
+            requests.exceptions.ConnectionError,
+            requests.exceptions.Timeout,
+        )
+
+        if DATABASE_SERVICE in config[SERVICES_TO_INSTALL]:
+            # Using the etcd CA cert as it's the same, and the permissions
+            # to its directory are more permissive than postgres'
+            ca_path = ETCD_CA_PATH
+        else:
+            ca_path = constants.POSTGRESQL_CA_CERT_PATH
+
+        try:
+            return requests.get(
+                url,
+                verify=ca_path,
+                timeout=5,
+            ).json()
+        except dead_node_exceptions as err:
+            logger.warning(
+                'Failed to get status of {target_type} node from {url}. '
+                'Error was: {err}'.format(
+                    target_type=target_type,
+                    address=address,
+                    url=url,
+                    err=err,
+                )
+            )
+            return {}
+
+    def _process_node_status(self, address, status, etcd_status,
+                             sync_replica=False, master=False):
+        node = {
+            'node_ip': address,
+            'alive': False,
+            'errors': [],
+            'raw_status': status,
+        }
+
+        if status:
+            xlog = status.get('xlog', {})
+            if master:
+                node['log_location'] = xlog.get('location')
+                node['state'] = 'leader'
+            elif sync_replica:
+                node['log_location'] = xlog.get('replayed_location')
+                node['state'] = 'sync_replica'
+            else:
+                node['log_location'] = xlog.get('replayed_location')
+                node['state'] = 'async_replica'
+
+            if status.get('state') == 'running':
+                node['alive'] = True
+            else:
+                node['errors'].append('Node not running')
+
+            node['timeline'] = status.get('timeline')
+        else:
+            node['state'] = 'dead'
+            node['errors'].append('Could not retrieve DB status')
+
+        if etcd_status:
+            node['etcd_state'] = etcd_status['state']
+        else:
+            node['etcd_state'] = 'dead'
+            node['errors'].append('Could not retrieve etcd status')
+
+        return node
+
+    def get_cluster_status(self):
+        # Start healthy, and let the status get worse as we find problems
+        status, exit_code = self._update_db_cluster_status(None, 'healthy')
```

Comment:
handling 2 values everywhere is very uncomfortable. I think I'd do:
```python
status = self.status['healthy']
if blabla:
 status |= self.status['down']
elif blabla:
 status |= self.status['degraded']
return status
```
(or even constants rather than dict items)
(especially since we don't actually need the string itself anywhere!)

that way we also lose the `_update_db_cluster_status` method entirely

and that way we possibly can divorce this method (which is on the "talk to the cluster" abstraction level) from human-readable-output-level, and from process-exit-code-level, both of which I feel should live in let's call it the presentation layer, not the logic layer

Reply:
Updated per further discussion in slack.